Remove logger.debug statements in DTensor dispatch

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #162337
* __->__ #161596
* #162218
* #162220
* #162219
* #161692
* #161634
* #161633
* #161595
* #161591

These seem to have been costing us 5-10 usec per detach (out of ~~95 usec total).  If they need to ship let's talk about requirements and how we can make this more efficient given that we would prefer if an entire DTensor op could finish in 10 usec.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

Differential Revision: [D81530106](https://our.internmc.facebook.com/intern/diff/D81530106)

swolchok committed:
Remove logger.debug statements in DTensor dispatch

These seem to have been costing us 5-10 usec per detach (out of ~~95 usec total).  If they need to ship let's talk about requirements and how we can make this more efficient given that we would prefer if an entire DTensor op could finish in 10 usec.

[ghstack-poisoned]

ezyang reviewed (approved):
I am grudgingly approving this. Is there any way we can have a debug statement along these lines without nuking perf?

swolchok committed:
Update on "Remove logger.debug statements in DTensor dispatch"


These seem to have been costing us 5-10 usec per detach (out of ~~95 usec total).  If they need to ship let's talk about requirements and how we can make this more efficient given that we would prefer if an entire DTensor op could finish in 10 usec.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

Skylion007 commented on torch/distributed/tensor/_dispatch.py:
```diff
@@ -144,7 +145,6 @@ def dispatch(
 
         # extract local tensor and sharding infos to a OpInfo
         op_info = self.unwrap_to_op_info(op_call, args, kwargs)
-        logger.debug("Dispatching op_call: %s", op_info.schema)
```
These should not have that much overhead, do they really?

swolchok replied:
> do they really

My measurements are noisy but per the summary it looked to me like they cost 5-10 usec per dispatch. timeit says the cost for a single trivial disabled log is most of a microsecond (result is total seconds to run the benchmark and it's 1 million iterations by default), and this was multiple logs that also pass arguments.

```
>>> timeit.timeit(stmt="logger.debug('hello')", setup="import logging; logger = logging.getLogger('test')")
0.8429189168382436
```

Skylion007 reviewed (approved):
Massive loss for logging. How unfortunate the overhead is so high even with lazy string substitution.

Skylion007 commented:
> > I am grudgingly approving this. Is there any way we can have a debug statement along these lines without nuking perf?
> 
> `if DEBUG: print("whatever")` comes to mind I guess?

```python
if logger.isEnabledFor(logging.debug):
```
is what you want.

swolchok committed:
rebase on "Remove logger.debug statements in DTensor dispatch"


These seem to have been costing us 5-10 usec per detach (out of ~~95 usec total).  If they need to ship let's talk about requirements and how we can make this more efficient given that we would prefer if an entire DTensor op could finish in 10 usec.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

swolchok commented:
> > > I am grudgingly approving this. Is there any way we can have a debug statement along these lines without nuking perf?
> > 
> > 
> > `if DEBUG: print("whatever")` comes to mind I guess?
> 
> ```python
> if logger.isEnabledFor(logging.debug):
>    pass
> ```
> 
> is what you want.

this check is still most of a microsecond:

```
timeit.timeit(stmt="logger.isEnabledFor(logging.DEBUG)", setup="import logging; logger = logging.getLogger('test')")
0.7279661330394447
```

Python is slow. :(

swolchok committed:
Update on "Remove logger.debug statements in DTensor dispatch"


These seem to have been costing us 5-10 usec per detach (out of ~~95 usec total).  If they need to ship let's talk about requirements and how we can make this more efficient given that we would prefer if an entire DTensor op could finish in 10 usec.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

swolchok commented:
@swolchok has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D81530106).

Skylion007 commented:
> > > > I am grudgingly approving this. Is there any way we can have a debug statement along these lines without nuking perf?
> > > 
> > > 
> > > `if DEBUG: print("whatever")` comes to mind I guess?
> > 
> > 
> > ```python
> > if logger.isEnabledFor(logging.debug):
> >    pass
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > is what you want.
> 
> this check is still most of a microsecond:
> 
> ```
> timeit.timeit(stmt="logger.isEnabledFor(logging.DEBUG)", setup="import logging; logger = logging.getLogger('test')")
> 0.7279661330394447
> ```
> 
> Python is slow. :(

True, but we can cache it for both calls if a single call is a fast enough speedup.

swolchok committed:
Update on "Remove logger.debug statements in DTensor dispatch"


These seem to have been costing us 5-10 usec per detach (out of ~~95 usec total).  If they need to ship let's talk about requirements and how we can make this more efficient given that we would prefer if an entire DTensor op could finish in 10 usec.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

Differential Revision: [D81530106](https://our.internmc.facebook.com/intern/diff/D81530106)

[ghstack-poisoned]

swolchok committed:
Update on "Remove logger.debug statements in DTensor dispatch"


These seem to have been costing us 5-10 usec per detach (out of ~~95 usec total).  If they need to ship let's talk about requirements and how we can make this more efficient given that we would prefer if an entire DTensor op could finish in 10 usec.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

Differential Revision: [D81530106](https://our.internmc.facebook.com/intern/diff/D81530106)

[ghstack-poisoned]

ezyang commented:
I wouldn't over-index on the test case for the logs; it is likely there to make sure people don't accidentally remove the log, as opposed to saying YOU MUST NOT REMOVE THE LOG (but as usual, the history is the truth)

swolchok committed:
Update on "Remove logger.debug statements in DTensor dispatch"


These seem to have been costing us 5-10 usec per detach (out of ~~95 usec total).  If they need to ship let's talk about requirements and how we can make this more efficient given that we would prefer if an entire DTensor op could finish in 10 usec.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

Differential Revision: [D81530106](https://our.internmc.facebook.com/intern/diff/D81530106)

[ghstack-poisoned]

swolchok committed:
Update on "Remove logger.debug statements in DTensor dispatch"


These seem to have been costing us 5-10 usec per detach (out of ~~95 usec total).  If they need to ship let's talk about requirements and how we can make this more efficient given that we would prefer if an entire DTensor op could finish in 10 usec.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

Differential Revision: [D81530106](https://our.internmc.facebook.com/intern/diff/D81530106)

[ghstack-poisoned]
